File: rsocket/rsocket-websocket-server/src/main/java/com/example/RSocketController.java
package com.example;

import java.time.LocalTime;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.rsocket.annotation.ConnectMapping;
import org.springframework.stereotype.Controller;

import reactor.core.publisher.Mono;

@Controller
public class RSocketController {
    private static final Logger LOGGER = LoggerFactory.getLogger(RSocketController.class);

    @ConnectMapping
    public Mono<Void> connect() {
        LOGGER.info("connect");
        return Mono.empty();
    }

    @MessageMapping("request_response")
    public Mono<HelloResponse> hello(HelloRequest request) {
        final HelloResponse response = new HelloResponse("Hello, %s!".formatted(request.name()),
                                                         LocalTime.now());
        return Mono.just(response);
    }

    public record HelloRequest(String name) {}

    public record HelloResponse(String message, LocalTime time) {}
}

File: kotlin/coroutines/rsocket/rsocket-server/src/main/kotlin/com/example/RSocketController.kt
package com.example

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.stereotype.Controller
import java.time.Instant

data class GreetingRequest(val name: String)
data class GreetingResponse(val message: String)

@Controller
class RSocketController {

    @MessageMapping("greetings")
    fun greet(request: GreetingRequest): Flow<GreetingResponse> = flow {
        while (true) {
            emit(GreetingResponse("${request.name} @ ${Instant.now()}"))
            delay(1000)
        }
    }
}
